File: 15-3sum/15-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Set<List<Integer>> totalSet = new HashSet<>();
        Arrays.sort(nums);
        int j, k;
        for ( int i = 0; i < nums.length - 2; i++){
            j = i + 1;
            k = nums.length - 1;
            while(j < k){
                if(nums[i] + nums[j] + nums[k] > 0)
                    k--;
                else if(nums[i] + nums[j] + nums[k] < 0)
                    j++;
                else{
                    totalSet.add(Arrays.asList(nums[i], nums[j], nums[k]));
                    j++;
                    k--;
                }
            }
        }
        return new ArrayList<>(totalSet);
    }
}

File: 15-3sum/15-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res = []
        for i in range(len(nums)):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            self.addToList(res, -nums[i], i + 1, nums)
        return res
            
    def addToList(self, res: List[int], targetSum: int, left: int, nums: List[int]):
        right = len(nums) - 1

        while (left < right):
            if nums[left] + nums[right] == targetSum:
                res.append([-targetSum, nums[left], nums[right]])
                left+=1
                right-=1
                while left < right and nums[left] == nums[left - 1]:
                    left+=1
                while left < right and nums[right] == nums[right + 1]:
                    right -= 1
            elif nums[left] + nums[right] < targetSum:
                left += 1
            else:
                right -= 1
            